src/ppx/metrics.cpp

```diff
@@ -0,0 +1,221 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/metrics.h"
+
+namespace ppx {
+namespace metrics {
+
+////////////////////////////////////////////////////////////////////////////////
+
+MetricGauge::MetricGauge(const MetricMetadata& metadata)
+    : mMetadata(metadata), mAccumulatedValue(0)
+{
+    memset(&mBasicStatistics, 0, sizeof(mBasicStatistics));
+    mBasicStatistics.min = std::numeric_limits<double>::max();
+    mBasicStatistics.max = std::numeric_limits<double>::min();
+}
+
+MetricGauge::~MetricGauge()
+{
+}
+
+void MetricGauge::RecordEntry(double seconds, double value)
+{
+    PPX_ASSERT_MSG(seconds >= 0.0, "The entries' seconds must always be positive");
+    PPX_ASSERT_MSG(mTimeSeries.size() == 0 || seconds > mTimeSeries.back().seconds, "The entries' seconds must form a stricly increasing function");
+
+    TimeSeriesEntry entry;
+    entry.seconds = seconds;
+    entry.value   = value;
+    mTimeSeries.push_back(entry);
+
+    UpdateBasicStatistics(seconds, value);
+}
+
+size_t MetricGauge::GetEntriesCount() const
+{
+    return mTimeSeries.size();
+}
+
+void MetricGauge::GetEntry(size_t index, double* seconds, double* value) const
```

Comment:
Should we validate that index is in range?

Reply:
I added an assert and a unit test to check the assert triggers.